Google.Api.Generator/ProtoUtils/ProtoExtensions.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Protobuf.Reflection;
+using System.Linq;
+using System.Reflection;
+
+namespace Google.Api.Generator.ProtoUtils
+{
+    static class ProtoExtensions
+    {
+        public static string CSharpNamespace(this FileDescriptor desc)
+        {
+            // There is no way of reading the C# namespace without using reflection.
```

Comment:
Ouch, did I know about this? Maybe I did. Ick - sorry about that.

Reply:
Yes, we discussed this a few months ago. This is fine for the time being - it's nicely contained, this is the only place this is required.
If/when protobuf is changed I'll update this.